lib/web_ui/test/embedder_test.dart

```diff
@@ -109,10 +87,174 @@ void testMain() {
     expect(style, isNotNull);
     expect(style.opacity, '0');
   }, skip: browserEngine != BrowserEngine.firefox);
+
+  group('Shadow root and styles', () {
+    late FlutterViewEmbedder embedder;
+
+    setUp(() {
+      embedder = FlutterViewEmbedder();
+    });
+
+    tearDown(() {
+      embedder.glassPaneElement.remove();
+    });
+
+    test('throws when shadowDom is not available', () {
+      final dynamic oldAttachShadow = attachShadow;
+      expect(oldAttachShadow, isNotNull);
+
+      attachShadow = null; // Break ShadowDOM
+
+      expect(() => FlutterViewEmbedder(), throwsUnsupportedError);
+      attachShadow = oldAttachShadow; // Restore ShadowDOM
+    });
+
+    test('Initializes and attaches a shadow root', () {
+      expect(domInstanceOfString(embedder.glassPaneShadow, 'ShadowRoot'), isTrue);
+      expect(embedder.glassPaneShadow.host, embedder.glassPaneElement);
+      expect(embedder.glassPaneShadow, embedder.glassPaneElement.shadowRoot);
+
+      // The shadow root should be initialized with correct parameters.
+      expect(embedder.glassPaneShadow.mode, 'open');
+      if (browserEngine != BrowserEngine.firefox &&
+          browserEngine != BrowserEngine.webkit) {
+        // Older versions of Safari and Firefox don't support this flag yet.
+        // See: https://caniuse.com/mdn-api_shadowroot_delegatesfocus
+        expect(embedder.glassPaneShadow.delegatesFocus, isFalse);
+      }
+    });
+
+    test('Attaches a stylesheet to the shadow root', () {
+      final DomElement? style =
+          embedder.glassPaneShadow.querySelector('#flt-internals-stylesheet');
+
+      expect(style, isNotNull);
+      expect(style!.tagName, equalsIgnoringCase('style'));
+    });
+
+    test('(Self-test) hasCssRule can extract rules', () {
+      final DomElement? style =
+          embedder.glassPaneShadow.querySelector('#flt-internals-stylesheet');
+
+      final bool hasRule = hasCssRule(style,
+          selector: '.flt-text-editing::placeholder',
+          declaration: 'opacity: 0');
+
+      final bool hasFakeRule = hasCssRule(style,
+          selector: 'input::selection', declaration: 'color: #fabada;');
+
+      expect(hasRule, isTrue);
+      expect(hasFakeRule, isFalse);
+    });
+
+    test('Attaches outrageous text styles to flt-scene-host', () {
```

Comment:
I'd move these tests to a separate `global_styles_test.dart` where we assert the things that must be contained in the sheet, then in the embedder_test leave only the ones that assert that a sheet was added at the right spot (`hasCssRule` shouldn't live in `embedder_test.dart`, but in `global_stles_test.dart`, IMO)

Reply:
Done